### PR TITLE
Product Visibility Taxonomies

### PR DESCRIPTION
**Pull request by mikejolley**
For #12524 

Meta queries are expensive, and in most cases we run 2 (joins and all) for most product queries;

- Visibility
- Stock status

To improve performance I’ve toyed with the idea of using a single taxonomy and it appears to work quite well.

Essentially this introduces a new private taxonomy called product visibility. It has terms:

- exclude-from-search
- exclude-from-catalog
- featured
- outofstock

Then these are used in place of meta queries.

I tested with 8000 products.

In 2.6, first load was ~9 seconds.

On this branch, it was ~0.5 seconds.

**Pros and cons**
+ Faster catalog queries in general.
+ Minimal impact on codebase due to CRUD.
+ Relatively simple upgrade routine.
- Some BW compatibility lost if using meta_query directly.
- Lose ability to "sort" by things we use terms for.

Hows it look @justinshreve @mattyza @claudiosanches @coderkevin? Spot any other issues with this approach?

Pining @gedex and @franticpsyx who may also be curious.

### CODE REVIEW

**Comment by franticpsyx** [comment]
I like the idea a lot. Don't see any obvious cons.

**Comment by mattyza** [comment]
LGTM, @mikejolley.

> Lose ability to "sort" by things we use terms for.

As in, we won't be able to sort by category? Would you mind elaborating on this point a bit?

Also, a heads up to @lucasstark as this may affect Catalog Visibility.

**Comment by mikejolley** [comment]
@mattyza So in this example, you would not be able to sort by 'featured'. Not that it's very good for sorting, but it was sortable via column header in 2.6.x

**Comment by gedex** [comment]
Nice works @mikejolley. I like your approach on using taxonomy.

> In 2.6, first load was ~9 seconds.

Which part of the page?

**Comment by mikejolley** [comment]
@gedex The main query's SQL COUNT ROWS part was 9 secs.

**Comment by gedex** [comment]
> The main query's SQL COUNT ROWS part was 9 secs

Is there any particular page that hits that query? Report? or Products screen maybe?

**Comment by mikejolley** [comment]
It was the shop page on frontend :)

File: includes/admin/class-wc-admin-post-types.php [hunk 1]
```diff
@@ -1876,27 +1869,23 @@ public function hidden_meta_boxes( $hidden, $screen ) {
...
-		$visibility_options = apply_filters( 'woocommerce_product_visibility_options', array(
-			'visible' => __( 'Catalog/search', 'woocommerce' ),
-			'catalog' => __( 'Catalog', 'woocommerce' ),
-			'search'  => __( 'Search', 'woocommerce' ),
-			'hidden'  => __( 'Hidden', 'woocommerce' ),
-		) );
+		$thepostid          = $post->ID;
+		$product_object     = $thepostid ? wc_get_product( $thepostid ) : new WC_Product;
+		$current_visibility = $product_object->get_catalog_visibility();
+		$current_featured   = $product_object->get_featured() ? 'yes' : 'no';
```
Comment (justinshreve) [inline includes/admin/class-wc-admin-post-types.php]: We could use `wc_bool_to_string` here.

**Review by emdashcodes — APPROVED** [review]
PR LGTM. Good job on the performance improvements here :).

**Comment by spencerfinnell** [comment]
@mikejolley There is an SQL error here: https://github.com/woocommerce/woocommerce/blob/master/includes/wc-term-functions.php#L616

Joining the two with no space between causes: 

```
SELECT COUNT( DISTINCT posts.ID ) FROM wptests_posts as posts
		LEFT JOIN wptests_term_relationships AS rel ON posts.ID=rel.object_ID
		LEFT JOIN wptests_term_taxonomy AS tax USING( term_taxonomy_id )
		WHERE 	post_status = &#039;publish&#039;
		AND 	post_type 	= &#039;product&#039;
	AND term_taxonomy_id !=7AND term_id IN ( 156 )
```

See the missing space between `!=7` and `AND`

**Pull request merged**